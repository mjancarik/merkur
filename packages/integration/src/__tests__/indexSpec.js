import { loadStyleAssets, loadScriptAssets } from '../index';

describe('Merkur component', () => {
  let assets = [];
  let fakeAssetObjects = [];

  const fakeAssetObjectGenerator = () => {
    const fakeAssetObject = {
      onerror: jest.fn(),
      onload: jest.fn(),
      removeAttribute: (name) => delete fakeAssetObject[name],
      setAttribute: (name, value) => (fakeAssetObject[name] = value),
    };
    fakeAssetObjects.push(fakeAssetObject);

    return fakeAssetObject;
  };

  const resolveFakeAssets = () => {
    for (const fakeAsset of fakeAssetObjects) {
      fakeAsset.onload();
    }
  };

  beforeEach(() => {
    assets = [
      {
        name: 'widget.js',
        type: 'script',
        source: {
          es9:
            'http://localhost:4444/static/es9/widget.6961af42bfa3596bb147.js',
          es5:
            'http://localhost:4444/static/es5/widget.31c5090d8c961e43fade.js',
        },
        attr: {
          async: true,
          'custom-element': 'amp-fx-collection',
          defer: false,
        },
      },
      {
        name: 'polyfill1.js',
        test: 'return typeof window !== "undefined"',
        type: 'script',
        source: {
          es9:
            'http://localhost:4444/static/es9/polyfill.6961af42bfa3596bb147.js',
          es5:
            'http://localhost:4444/static/es5/polyfill.6961af42bfa3596bb147.js',
        },
      },
      {
        name: 'undefined.js',
        test: 'return typeof window !== "undefined"',
        type: 'script',
        source: {
          es9: undefined,
          es5: undefined,
        },
      },
      {
        name: 'polyfill2.js',
        test: 'return (function () {foo();})()',
        type: 'script',
        source:
          'http://localhost:4444/static/es5/polyfill.6961af42bfa3596bb147.js',
      },
      {
        type: 'inlineScript',
        source: 'alert();',
      },
      {
        name: 'widget.css',
        type: 'stylesheet',
        source:
          'http://localhost:4444/static/es9/widget.814e0cb568c7ddc0725d.css',
      },
      {
        name: 'undefined.css',
        type: 'stylesheet',
        source: undefined,
      },
      {
        type: 'inlineStyle',
        source: '.cssClass {Â margin-top: 5px; }',
      },
    ];

    fakeAssetObjects = [];
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  describe('loadStyleAssets() function', () => {
    it('should create style elements for style assets', (done) => {
      jest
        .spyOn(document, 'createElement')
        .mockImplementation(fakeAssetObjectGenerator);
      spyOn(document.head, 'appendChild');

      loadStyleAssets(assets).then(() => {
        expect(document.createElement).toHaveBeenCalledTimes(2);
        expect(document.createElement).toHaveBeenNthCalledWith(1, 'link');
        expect(document.createElement).toHaveBeenNthCalledWith(2, 'style');
        expect(document.head.appendChild).toHaveBeenCalledTimes(2);

        done();
      });

      resolveFakeAssets();

      expect(fakeAssetObjects).toMatchSnapshot();
    });

    it('should return promise that resolves after the styles are loaded', (done) => {
      jest
        .spyOn(document, 'createElement')
        .mockImplementation(fakeAssetObjectGenerator);
      spyOn(document.head, 'appendChild');

      const stylePromise = loadStyleAssets(assets).then(() => {
        expect(stylePromise).toBeInstanceOf(Promise);

        done();
      });

      resolveFakeAssets();
    });
  });

  describe('loadScriptAssets() function', () => {
    it('should create script elements for script assets', (done) => {
      jest
        .spyOn(document, 'createElement')
        .mockImplementation(fakeAssetObjectGenerator);
      spyOn(document.head, 'appendChild');

      loadScriptAssets(assets).then(() => {
        expect(document.createElement).toHaveBeenCalledTimes(3);
        expect(document.head.appendChild).toHaveBeenCalledTimes(3);

        done();
      });

      resolveFakeAssets();

      expect(fakeAssetObjects).toMatchSnapshot();
    });

    it('should return promise that resolves after the scripts are loaded', (done) => {
      jest
        .spyOn(document, 'createElement')
        .mockImplementation(fakeAssetObjectGenerator);
      spyOn(document.head, 'appendChild');

      const scriptsPromise = loadScriptAssets(assets).then(() => {
        expect(scriptsPromise).toBeInstanceOf(Promise);

        done();
      });

      resolveFakeAssets();
    });
  });
});
