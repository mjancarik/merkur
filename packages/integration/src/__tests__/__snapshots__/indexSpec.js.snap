// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Merkur component loadScriptAssets() function should create script elements for script assets 1`] = `
Array [
  Object {
    "async": "",
    "custom-element": "amp-fx-collection",
    "onerror": [Function],
    "onload": [Function],
    "removeAttribute": [Function],
    "setAttribute": [Function],
    "src": "http://localhost:4444/static/es9/widget.6961af42bfa3596bb147.js",
  },
  Object {
    "defer": true,
    "onerror": [Function],
    "onload": [Function],
    "removeAttribute": [Function],
    "setAttribute": [Function],
    "src": "http://localhost:4444/static/es5/polyfill.6961af42bfa3596bb147.js",
  },
  Object {
    "onerror": [MockFunction],
    "onload": [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
    "removeAttribute": [Function],
    "setAttribute": [Function],
    "text": "alert();",
  },
]
`;

exports[`Merkur component loadStyleAssets() function should create style elements for style assets 1`] = `
Array [
  Object {
    "href": "http://localhost:4444/static/es9/widget.814e0cb568c7ddc0725d.css",
    "onerror": [Function],
    "onload": [Function],
    "removeAttribute": [Function],
    "setAttribute": [Function],
  },
  Object {
    "innerHTML": ".cssClass {Â margin-top: 5px; }",
    "onerror": [MockFunction],
    "onload": [MockFunction] {
      "calls": Array [
        Array [],
      ],
      "results": Array [
        Object {
          "type": "return",
          "value": undefined,
        },
      ],
    },
    "removeAttribute": [Function],
    "setAttribute": [Function],
  },
]
`;
